src/foam/nanos/approval/ApprovalRequestNotificationDAO.java

```diff
@@ -0,0 +1,98 @@
+import foam.nanos.approval.ApprovalStatus;
+import foam.nanos.notification.Notification;
+import net.nanopay.meter.compliance.ComplianceApprovalRequest;
+import static foam.mlang.MLang.*;
+
+public class ApprovalRequestNotificationDAO
+extends ProxyDAO {
+  public ApprovalRequestNotificationDAO(X x,DAO delegate) {
+    setX(x);
+    setDelegate(delegate);
+  }
+
+  @Override
+  public FObject put_(X x, FObject obj) {
+    ApprovalRequest old = (ApprovalRequest) getDelegate().find(obj);
+    ApprovalRequest ret = (ApprovalRequest) getDelegate().put_(x, obj);
+    String causeDAO = "";
+    String notificationType = "";
+    String notificationBody = "";
+
+
+    if ( ret instanceof ComplianceApprovalRequest ) {
+      ComplianceApprovalRequest complianceApprovalRequest = (ComplianceApprovalRequest) ret;
+      causeDAO = complianceApprovalRequest.getCauseDaoKey();
+      notificationType = "approval request for reference id: " + ret.getRefObj() + " cause: " + causeDAO;
+      notificationBody = "New approval was requested for reference id: " + ret.getRefObj() + " cause: " + causeDAO;
+    } else if ( old != null
+      && ret.getStatus() != old.getStatus()
+      && ( ret.getStatus() == ApprovalStatus.APPROVED || ret.getStatus() == ApprovalStatus.REJECTED )
+      && ret.getLastModifiedBy() != ret.getApprover()
+    ) {
+      notificationType = "approval request updated for reference id: " + ret.getRefObj();
+      notificationBody = new StringBuilder("Approval request for reference id: ")
+        .append(ret.getRefObj())
+        .append(" sent to you has been approved by: ")
+        .append(ret.getLastModifiedBy())
+        .toString();
+    } else if ( old == null ) {
+      notificationType = "approval request for reference id: " + ret.getRefObj();
+      notificationBody = "New approval was requested for reference id: " + ret.getRefObj();
+    } else {
+      return ret;
+    }
+
+    Notification notification = new Notification.Builder(x)
+      .setUserId(ret.getApprover())
+      .setNotificationType(notificationType)
+      .setEmailIsEnabled(true)
+      .setBody(notificationBody)
+      .build();
+    //notification.setEmailName("future email template name"); !!! PROPER WAY TO SET EMAIL TEMPLATE (when it is done) !!!
+    //notification.setEmailArgs(MAP_GOES_HERE); !!! PROPER WAY TO SET EMAIL ARGS FOR TEMPLATE !!!
+    ret.findApprover(x).doNotify(x, notification);
+    return ret;
+  }
+
+  @Override
+  public FObject remove_(X x, FObject obj) {
+    ApprovalRequest ret = (ApprovalRequest) getDelegate().remove_(x, obj);
+    ApprovalRequest fulfilled = (ApprovalRequest) getDelegate().find(AND(
+        OR(
+          EQ(ApprovalRequest.STATUS, ApprovalStatus.REJECTED),
+          EQ(ApprovalRequest.STATUS, ApprovalStatus.APPROVED)
+        ),
+        EQ(ApprovalRequest.CREATED_BY, ret.getCreatedBy()),
+        EQ(ApprovalRequest.OBJ_ID, ret.getObjId()),
+        EQ(ApprovalRequest.APPROVABLE_HASH_KEY, ret.getApprovableHashKey())
```

Comment:
@noodlemoodle don't need approvable hash key for remove since its just the object itself, no check for properties or anything

Reply:
but this is for removing the approvalrequest no? not operations.remove @tharmaman

Reply:
My bad. But why does the Approvable_Hash_Key have to exist here? Isn't the object id sufficient? Obj_Id either refers to the Approvable or the object itself. @noodlemoodle

Reply:
hmmmm make sense but I think for approvable the obj_id is just the daokey + id of obj being updated now, so there could be multiple approvalrequests for different update operations for the same "objId" right? so its not guaranteed the correct "fulfilled request" will be returned  @tharmaman

Reply:
Yeah so in the rework i did, approvables now use GUID and the GUID is stored as the obj_id and a new property lookup_id is on approvable which has that hash. So I'll remove the Approvable HAsh Key in my PR because we don't even set it in the approvableAwareDAO

Reply:
oh true ok makes sense i should look thru the rework first @tharmaman